### PR TITLE
Add an environment variable that determines whether pinned memory usage should be restricted.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [X] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
Memory tuning - add an environment flag restrict pinned memory usage.

#### Additional information
Buffers
Executor
NVJPEG

### Tests
- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2337

### CODE REVIEW

**Review by JanuszL — APPROVED** [review]

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h [hunk 1]
```diff
@@ -241,7 +243,7 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
       }
       if (host_memory_padding_jpeg2k > 0) {
         nvjpeg_memory::AddBuffer<mm::memory_kind::pinned>(nvjpeg2k_thread_id,
-                                 host_memory_padding_jpeg2k);
+                                                          host_memory_padding_jpeg2k);
```
Comment (jantonguirao) [inline dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:246]: Can't we make this one regular host memory as well? I'd leave a note if there is a good reason why we didn't change that one

File: dali/operators/decoder/nvjpeg/nvjpeg_memory.cc [hunk 2]
```diff
@@ -311,6 +332,32 @@ static int PinnedNew(void **ptr, size_t size, unsigned int flags) {
...
+  }
+  // this function should not throw, but return a proper result
+  try {
+    if (RestrictPinnedMemUsage()) {
+      *ptr = GetBuffer<mm::memory_kind::host>(std::this_thread::get_id(), size);
+      return *ptr != nullptr ? cudaSuccess : cudaErrorMemoryAllocation;
+    } else {
+      *ptr = GetBuffer<mm::memory_kind::pinned>(std::this_thread::get_id(), size);
+      return *ptr != nullptr ? cudaSuccess : cudaErrorMemoryAllocation;
+    }
```
Comment (jantonguirao) [inline dali/operators/decoder/nvjpeg/nvjpeg_memory.cc:349]:
```suggestion
    if (RestrictPinnedMemUsage()) {
      *ptr = GetBuffer<mm::memory_kind::host>(std::this_thread::get_id(), size);
    } else {
      *ptr = GetBuffer<mm::memory_kind::pinned>(std::this_thread::get_id(), size);
    }
    return *ptr != nullptr ? cudaSuccess : cudaErrorMemoryAllocation;
```
Just a suggestion

File: dali/pipeline/data/buffer.h [hunk 3]
```diff
@@ -37,6 +37,7 @@ class CPUBackend;
 
 DLL_PUBLIC shared_ptr<uint8_t> AllocBuffer(size_t bytes, bool pinned, GPUBackend *);
 DLL_PUBLIC shared_ptr<uint8_t> AllocBuffer(size_t bytes, bool pinned, CPUBackend *);
+DLL_PUBLIC bool RestrictPinnedMemUsage();
```
Comment (jantonguirao) [inline dali/pipeline/data/buffer.h:40]: a small doxygen here would help, I think

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h (hunk 1 above)
Comment (mzient) [reply dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:246]: Nope, nvjpeg2k crashed with regular host buffer.

File: dali/operators/decoder/nvjpeg/nvjpeg_memory.cc [hunk 4]
```diff
@@ -87,6 +88,7 @@ struct NVJpegMem {
 
   static NVJpegMem &instance() {
     // ensure proper destruction order
+    (void)mm::GetDefaultResource<mm::memory_kind::host>();
```
Comment (JanuszL) [inline dali/operators/decoder/nvjpeg/nvjpeg_memory.cc:91]: Side question - these resources still use the global, common pool but provides separate stats regarding allocations, right?

File: dali/python/backend_impl.cc [hunk 5]
```diff
@@ -183,6 +183,8 @@ void FillTensorFromDlPack(py::capsule capsule, SourceDataType<SrcBackend> *batch
     batch->set_pinned(true);
   } else if (dl_tensor.device.device_type == kDLCPU) {
     batch->set_device_id(-1);
+    if (RestrictPinnedMemUsage())
+      batch->set_pinned(false);
```
Comment (JanuszL) [inline dali/python/backend_impl.cc:187]:
```suggestion
    batch->set_pinned(false);
```
`batch` just wraps the memory provided by DLPack. So it if it kDLCPU it is not be pinned.

File: dali/test/python/test_pipeline.py [hunk 6]
```diff
@@ -1430,7 +1433,7 @@ def obtain_reader_meta(iters = 3, **kvargs):
             pipe.set_outputs(*out)
         pipe.build()
         for _ in range(iters):
-            pipe.run()
+            outs = pipe.run()
```
Comment (JanuszL) [inline dali/test/python/test_pipeline.py:1436]: ?

File: dali/python/backend_impl.cc (hunk 5 above)
Comment (mzient) [reply dali/python/backend_impl.cc:187]: I didn't want to slow it down, so if it was pinned, I don't want to make it unconditionally non-pinned.

File: dali/test/python/test_pipeline.py (hunk 6 above)
Comment (mzient) [reply dali/test/python/test_pipeline.py:1436]: Leftover from experiments.

File: dali/python/backend_impl.cc (hunk 5 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc:187]: It never was. I think it is just a bug.
kDLCUDAHost - pinned
kDLCPU - not pinned
And:
```
batch->ShareData(shared_ptr<void>(dl_tensor.data,
                                    [dlm_tensor_ptr = move(dlm_tensor_ptr)](void*) {}),
                                    bytes, typed_shape, dali_type.id());```
just accepts the data from the outside world, and all we know about it is inside `device_type`.
I think the bug here comes from the fact that our tensors are pinned by default what can be easily missed.

Comment (JanuszL) [reply dali/python/backend_impl.cc:187]: But we can leave it as it is.

File: dali/pipeline/data/buffer.h (hunk 3 above)
Comment (mzient) [reply dali/pipeline/data/buffer.h:44]: Done.

**Review by JanuszL — APPROVED** [review]

File: dali/python/backend_impl.cc (hunk 5 above)
Comment (mzient) [reply dali/python/backend_impl.cc:187]: OK, I'll change it.

**Review by JanuszL — APPROVED** [review]

**Pull request merged**